experimental/FTheoryTools/auxiliary.jl

```diff
@@ -0,0 +1,319 @@
+
+    # Construct the incidence matrix for the maximal cones of the ambient space
+    ambient_space_max_cones = []
+    for i in 1:nrows(base_cones)
+        push!(ambient_space_max_cones, [k for k in hcat([b for b in base_cones[i,:]], [1 1 0])])
+        push!(ambient_space_max_cones, [k for k in hcat([b for b in base_cones[i,:]], [1 0 1])])
+        push!(ambient_space_max_cones, [k for k in hcat([b for b in base_cones[i,:]], [0 1 1])])
+    end
+    ambient_space_max_cones = IncidenceMatrix(vcat(ambient_space_max_cones...))
+
+    # Construct and return the ambient space
+    toric_ambient_space = normal_toric_variety(PolyhedralFan(ambient_space_rays, ambient_space_max_cones; non_redundant = true))
+    set_coordinate_names(toric_ambient_space, vcat([string(k) for k in gens(cox_ring(base))], ["x", "y", "z"]))
+    return toric_ambient_space
+
+end
+
+
+################################################################
+# 3: Construct the Weierstrass polynomial
+################################################################
+
+function _weierstrass_sections(base::AbstractNormalToricVariety)
+    f = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base)^4)])
+    g = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base)^6)])
+    return [f, g]
+end
+
+function _weierstrass_polynomial(base::AbstractNormalToricVariety, S::MPolyDecRing{fmpq, FmpqMPolyRing})
+    (f, g) = _weierstrass_sections(base)
+    return _weierstrass_polynomial(S, f, g)
+end
+
+function _weierstrass_polynomial(f::MPolyElem{fmpq}, g::MPolyElem{fmpq}, S::MPolyDecRing{fmpq, FmpqMPolyRing})
+    x = gens(S)[length(gens(S))-2]
+    y = gens(S)[length(gens(S))-1]
+    z = gens(S)[length(gens(S))]
+    ring_map = hom(parent(f), S, [gens(S)[i] for i in 1:length(gens(S))-3])
+    return x^3 - y^2 + ring_map(f)*x*z^4 + ring_map(g)*z^6
+end
+
+
+################################################################
+# 4: Construct the Tate polynomial
+################################################################
+
+function _tate_sections(base::AbstractNormalToricVariety)
+    a1 = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base))])
+    a2 = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base)^2)])
+    a3 = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base)^3)])
+    a4 = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base)^4)])
+    a6 = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base)^6)])
+    return [a1, a2, a3, a4, a6]
+end
+
+function _tate_polynomial(base::AbstractNormalToricVariety, S::MPolyDecRing{fmpq, FmpqMPolyRing})
+    (a1, a2, a3, a4, a6) = _tate_sections(base)
+    return _tate_polynomial(S, [a1, a2, a3, a4, a6])
+end
+
+function _tate_polynomial(ais::Vector{<:MPolyElem{fmpq}}, S::MPolyDecRing{fmpq, FmpqMPolyRing})
+    x = gens(S)[length(gens(S))-2]
```

Comment:
Don't use `gens(S)` as a cheap disposable like you would do in GAP: in fact, for certain rings, *every* call to `gens` allocates a new fresh deepcopy of the generating set. This function calls `gens(S)` eight times... 

Besides, to get generator `i`, you can also use `gen(S, i)` which again is in general more efficient than `gens(S)[i]`